### DISCUSSION TITLE
✨ Use consistent emojis to indicate the PR/Issue type

### DISCUSSION THREAD

**Discussion by mbercx** [aiidateam/aiida-core#4589] [Ideas]
I recently found out that my MacBook pro's touch bar allows me to quickly and easily look for emojis. This discovery, in combination with some quarantine-induced bouts of mental instability, have made me _very_ giddy about using them pretty much everywhere (as anyone who's been reading my meeting notes recently can testify).

Then I saw the [executablebooks/sphinx-copybutton commit list](https://github.com/executablebooks/sphinx-copybutton/commits/master), and I got jealous of how pretty it looks.

So, maybe we can also start using these consistently for our PR's and commits? 🙏

On a more serious note, I think using these consistently can help with sorting the changes for the release changelogs, and it's shorter to add e.g. :books: than `DOCS:`, :wrench: than `FIX:` or  :sparkles: than `ENHANCEMENT:`.

**Comment by chrisjsewell** [comment]
I've been saying this for ages lol; who do you think implemented those emojis across the whole of the executablebooks organisation, this guy 🙋‍♂️

- https://github.com/executablebooks/meta/issues/74
- https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#commit-messages

**Reply by mbercx** [reply]
> I've been saying this for ages lol; who do you think implemented those emojis across the whole of the executablebooks organisation, this guy 🙋‍♂️

Hehe, I figured I would have your support. 😅  I was just curious what the others thought and if we could get the ⚽ rolling on this.

**Reply by chrisjsewell** [reply]
indeed and its already used for quantum mobile and all that organisation 😉 : https://github.com/marvel-nccr/quantum-mobile

**Reply by mbercx** [reply]
![amazing](https://user-images.githubusercontent.com/25607960/100221764-6cbdd480-2f19-11eb-8a70-e6cecfed45ba.gif)

**Reply by chrisjsewell** [reply]
![image](https://media.giphy.com/media/BSx6mzbW1ew7K/giphy.gif)

**Comment by greschd** [comment]
Well, for a bit of a contrarian view: I'm not really excited about the emoji business. While great for casual conversation, I wouldn't put it in commit messages.

Off the top of my head:
- can you `grep` for that?
- it's inconvenient for people that _don't_ have an emoji picker easily available
- I'm not sure if the  🔧 <-> `FIX` etc. translation is "universally accepted standard" - if it isn't, it seems a burden having to remember it for each project. At least at first glance, this is much less understandable to me
- Not sure about the accessibility of using emojis

That's just my gut reaction - but if everyone else loves it I wouldn't mind _all that much_ ¯\\\_(ツ)\_/¯

**Reply by chrisjsewell** [reply]
if you look on my conventions, you'll see that you use both the emoji and the text prefix.
Also executablebooks has a lot more maintainers/contributors that aiida, and none of them have had an issue with it

**Reply by chrisjsewell** [reply]
Also github recognises both the actual emoji or the GFM text representation of it (like `:spanner:`)

**Reply by greschd** [reply]
Yeah I guess both emoji + text would work.

> Also github recognises both the actual emoji or the GFM text representation of it (like :spanner:)

Doesn't that just invite inconsistency?

> Also executablebooks has a lot more maintainers/contributors that aiida

Still not a fan of the "larger community" argument, but a quick look tells me 59 vs. 61 contributors 🙃

**Reply by chrisjsewell** [reply]
I mean basically look at the commit history of aiida-core: https://github.com/aiidateam/aiida-core/commits/develop, and look at the commit history of jupyter-book: https://github.com/executablebooks/jupyter-book/commits/master, and you tell me what one is easier to digest

**Reply by chrisjsewell** [reply]
> Still not a fan of the "larger community" argument, but a quick look tells me 59 vs. 61 contributors

yeh and the executablebooks org was only created in april

**Reply by mbercx** [reply]
Thanks for your 2 cents, @greschd! Some thoughts:

> * can you `grep` for that?

Not sure what happens on GitHub if you add `:wrench:` (🔧 ) to your commit messages, but you _could_ grep for that. 😄 

EDIT: It works!

<img width="858" alt="Screenshot 2020-11-25 at 14 16 13" src="https://user-images.githubusercontent.com/25607960/100232605-db565e80-2f28-11eb-9070-6d1a1ca72225.png">

> * it's inconvenient for people that _don't_ have an emoji picker easily available

Well, you can use the [markdown emoji markup](https://gist.github.com/rxaviers/7360908) as well. This is what I do once I know an emoji and want to use it often. You do have to learn them, that's true.

> * I'm not sure if the  🔧 <-> `FIX` etc. translation is "universally accepted standard" - if it isn't, it seems a burden having to remember it for each project. At least at first glance, this is much less understandable to me

Is there a universally accepted standard for using `FIX`, `DOCS` etc? It's true that these identifiers are more widely used, however (for now? 💪). We can leave adding the emojis to PR's to the maintainers, who can find it useful when working on a release.

> * Not sure about the accessibility of using emojis

What do you mean with _accessibility_, here?

> That's just my gut reaction - but if everyone else loves it I wouldn't mind _all that much_ ¯\_(ツ)_/¯

I'm also not adamant that we _have_ to use it. 😅 Just was curious about people's opinion. 🤷

**Reply by chrisjsewell** [reply]
actually I guess you are only talking about jupyter-book has 61 contributors, there's 44 other repos in executablebooks

**Reply by greschd** [reply]
> actually I guess you are only talking about jupyter-book has 61 contributors, there's 44 other repos in executablebooks

I actually _really_ dislike this "appeal to authority" argument. Just because a community is larger, doesn't make its conventions "more correct". I can always find [a larger community](https://github.com/torvalds/linux) that does things differently.. that leads nowhere.

But yeah, I can kinda see how the commit log is easier to parse. Again, not fundamentally opposed if most people love it - just wanted to share my "meh" feeling about that.

To me it sometimes looks like a bit of an eye sore, but that's definitely on the "personal preference" side of things.

**Reply by chrisjsewell** [reply]
> I actually really dislike this "appeal to authority" argument. 

The point was not that it is more correct, it was in response to suggestions that it may be inaccessible/not user-friendly.

I just want to "robustly" defend it otherwise lol, because I do generally think its a lot better than what we currently have (our commit message are generally very good, but the titles are just all over the place)

**Reply by greschd** [reply]
>  response to suggestions that it may be inaccessible/not user-friendly.

Fair enough, point taken. 

>  its a lot better than what we currently have

Yeah agreed, _some_ standard should be used. Note there is an existing discussion here: https://github.com/aiidateam/aiida-core/issues/4065

**Reply by sphuber** [reply]
Here are my two cents. I am so far not really thrilled by the idea of the emoticon's, although I do think that Chris' motivation behind it of creating more consistency in commit messages is a very good one.

When it comes to git commit messages, I think the following points are very important:

* Commit messages should be consistently formatted. The title should quickly indicate the scope and the body should provide important background information and context, while detailing the design choices and any other valuable considerations.
* Commit messages should be easily created and read in the command line. Other text, such as documentation, can be built and optimized for reading in a browser, but commit messages are often created and read in a terminal and should be formatted for easy reading there.
* Commit messages should not be locked into to any one external platform by including information that can only be interpreted by that platform.

When we apply this to the proposal of using emoticon's, I come to the following conclusion:

* The emoticon's are trying to make the commit message titles more standardized and therefore easier to read/comprehend. This is a good aim
* However, I don't think emoticon's are the correct way of doing this. I think the meaning of emoticon's are not always necessarily clear and can actually be ambiguous
* It was proposed to have a text as a backup, but then why have the emoticon? What is the added value.
* Are the emoticon's always correctly displayed? Are they dependent on the browser/operation system/terminal in which the commit message is rendered? I think the inconsistency is undesirable and counter productive
* Will people know how to insert these emoticon's? Everyone can type text in their terminal and editor. Does the same hold for emoticon's? It was suggested to use the textual alternative, e.g. `:wrench:`, but this is Github specific and then one also needs to know all the shortcuts for all emoticons.

Finally:

> our commit message are generally very good, but the titles are just all over the place

Defining conventions is one thing, like I did a long time ago for [commit messages](
https://github.com/aiidateam/aiida-core/wiki/Git-style-requirements), but getting people to actually respect them is another question altogether. So if the idea is to create consistency in commit message (titles), we should take a well formalized convention and automatically lint messages and I would be more tempted to continue in the direction of #4065 which also will allow us to add linting.

**Reply by chrisjsewell** [reply]
some quick responses:

1. emojis are part of the utf8 standard convention, they are available on every modern platform and terminal and everyone knows how to use them.
2. The commit titles should be easy to scan not read, thats what the commit messages are for, its perfectly fine for them to be slightly ambiguous. No one sits there and reads all the commit titles unless they have to, the point is that you can very quickly look at a list of commits and identify their type; documentation, bug fixes, new features; then you can look at the full commit messages for greater context.
3. You can't lint commit messages, because they are written after the CI, generally within the GitHub web interface

**Reply by sphuber** [reply]
>  emojis are part of the utf8 standard convention, they are available on every modern platform and terminal and everyone knows how to use them.

I might be a ludite, but I have no clue how to insert an emoticon from my terminal.

>  The commit titles should be easy to scan not read, thats what the commit messages are for, its perfectly fine for them to be slightly ambiguous. No one sits there and reads all the commit titles unless they have to, the point is that you can very quickly look at a list of commits and identify their type; documentation, bug fixes, new features; then you can look at the full commit messages for greater context.

I agree that the titles should be quick to scan, but the "type" is not the only important information. I would argue that the scope or subject of the commit is even more important, i.e. what part of the code is affected, e.g., `Engine`, `CalcJob`, `WorkChain` or CLI, etc

>  You can't lint commit messages, because they are written after the CI, generally within the GitHub web interface

Commit messages are also written when you commit and so they can be linted using a commit post-hook.

**Reply by chrisjsewell** [reply]
> I might be a ludite, but I have no clue how to insert an emoticon from my terminal.

just copy/paste them into, if nothing else. But who knows, I'm not a ludite, I rarely use git from the terminal, when theres much better interfaces, like [vs code](https://code.visualstudio.com/docs/editor/github) or [Github desktop](https://desktop.github.com/) 😜 

>  what part of the code is affected, e.g., Engine, CalcJob, WorkChain or CLI, etc

this would rarely be singular though surely. It may be better to add a yaml-type tag to the top of the commit message, e.g. `tags: [engine, cli]`

> Commit messages are also written when you commit and so they can be linted using a commit post-hook.

if you can work out how to enforce this on the GitHub web interface, then I'm happy to stand corrected

**Reply by greschd** [reply]
> just copy/paste them into, if nothing else

🥴

On Windows there's `Win + .`, on Linux I haven't found a good consistent way of doing it. But copy/paste is IMO unacceptable, that's just a giant waste of time. Searching for the emojis in a picker is already tedious enough..

> theres much better interfaces, like vs code or Github desktop

Let's not pretend we can solve the age-old CLI vs GUI debate here..

**Reply by chrisjsewell** [reply]
well I guess we will have to agree to disagree
